dart/collision/CollisionGroup.cpp

```diff
@@ -0,0 +1,209 @@
+#include "dart/dynamics/Skeleton.h"
+
+namespace dart {
+namespace collision {
+
+//==============================================================================
+CollisionGroup::CollisionGroup(const CollisionDetectorPtr& collisionDetector)
+  : mCollisionDetector(collisionDetector)
+{
+  assert(mCollisionDetector);
+}
+
+//==============================================================================
+CollisionGroup::~CollisionGroup()
+{
+  // Do nothing
+}
+
+//==============================================================================
+CollisionDetector* CollisionGroup::getCollisionDetector()
+{
+  return mCollisionDetector.get();
+}
+
+//==============================================================================
+const CollisionDetector* CollisionGroup::getCollisionDetector() const
+{
+  return mCollisionDetector.get();
+}
+
+//==============================================================================
+void CollisionGroup::addShapeFrame(const dynamics::ShapeFrame* shapeFrame)
+{
+  if (!shapeFrame)
+    return;
+
+  if (hasShapeFrame(shapeFrame))
+    return;
+
+  auto collObj = mCollisionDetector->claimCollisionObject(shapeFrame);
+
+  notifyCollisionObjectAdded(collObj.get());
+
+  mShapeFrames.push_back(shapeFrame);
+  mCollisionObjects.push_back(collObj);
+}
+
+//==============================================================================
+void CollisionGroup::addShapeFrames(
+    const std::vector<const dynamics::ShapeFrame*>& shapeFrames)
+{
+  for (const auto& shapeFrame : shapeFrames)
+    addShapeFrame(shapeFrame);
+}
+
+//==============================================================================
+void CollisionGroup::addShapeFramesOf()
+{
+  // Do nothing
+}
+
+//==============================================================================
+void CollisionGroup::removeShapeFrame(const dynamics::ShapeFrame* shapeFrame)
+{
+  if (!shapeFrame)
+    return;
+
+  const auto search
+      = std::find_if(mShapeFrames.begin(), mShapeFrames.end(),
+                     [&](const dynamics::ShapeFrame* it)
+                     { return it == shapeFrame; });
+
+  if (mShapeFrames.end() == search)
+    return;
+
+  const size_t index = search - mShapeFrames.begin();
+
+  notifyCollisionObjectRemoved(mCollisionObjects[index].get());
+
+  mShapeFrames.erase(search);
+  mCollisionObjects.erase(mCollisionObjects.begin() + index);
+}
+
+//==============================================================================
+void CollisionGroup::removeShapeFrames(
+    const std::vector<const dynamics::ShapeFrame*>& shapeFrames)
+{
+  for (const auto& shapeFrame : shapeFrames)
+    removeShapeFrame(shapeFrame);
+}
+
+//==============================================================================
+void CollisionGroup::removeShapeFramesOf()
+{
+  // Do nothing
+}
+
+//==============================================================================
+void CollisionGroup::removeAllShapeFrames()
+{
+  notifyAllCollisionObjectsRemoved();
```

Comment:
I noticed that `notifyAllCollisionObjectsRemoved` and `notifyCollisionObjectRemoved` are called *before* the removal actually happens. That could lead to difficult-to-diagnose bugs if the callback manipulates the `CollisionGroup`. Is that intentional?